sidecar/register/registration.go

```diff
@@ -82,12 +83,14 @@ func (agent *RegistrationAgent) Stop() {
 
 func (agent *RegistrationAgent) register() {
 	for {
-		registeredInstance, err := agent.config.Client.Register(agent.config.ServiceInstance)
-		if err == nil {
-			go agent.renew(registeredInstance)
-			return
+		if agent.checkAppHealth(DefaultReregistrationDelay) {
```

Comment:
I'd like to suggest an alternative approach for combining the healthchecking / registration logic:

1. Leave the `RegistrationAgent` as is (before the PR).
2. Implement a `Healthchecker` object that continuously executes the configured healthchecks on the background.
3. The `Healthchecker` is either in healthy or unhealthy state at any given time, depending on the combined results of the healthchecks.
4. When the `Healthchecker` moves to healthy state, it immediately invokes `agent.Start()`. Similarly, when it moves to unhealthy state, it immediately invokes `agent.Stop()`.

What we gain:

1. Simpler state machine. Especially if we add more advanced stuff like different intervals for healthchecks, heartbeats, TTL, and grace period for failed healthchecks.
2. Faster response (registration/deregistration) to becoming healthy/unhealthy